File: packages/client/src/ffi/devices.ts
import { invoke } from "./invoke";

export type DeviceIds = Array<string>;

type Args = {
  add_devices: { deviceIds: DeviceIds };
  list_devices: never;
};

export const addDevices = async (deviceIds: DeviceIds) => {
  try {
    await invoke<Args, boolean>()("add_devices", { deviceIds });
  } catch (e) {
    console.error("[Device] Failed to add devices!", e);
  }
};

export const listDevices = async (): Promise<DeviceIds | void> => {
  try {
    const device_ids = await invoke<Args, DeviceIds>()(
      "list_devices",
      {} as never,
    );

    if (!device_ids) {
      throw new Error("Empty response");
    }

    return device_ids;
  } catch (e) {
    console.error("[Device] Failed to list devices!", e);
  }
};

File: crates/pyxis-app/src/handlers/devices.rs
use pyxis_shared::{database::ConfigDatabase, entities::devices::Device};
use tauri::State;

#[tauri::command]
pub fn add_devices(device_ids: Vec<String>, sync_db: State<ConfigDatabase>) -> Option<bool> {
    let devices = Device::new(None, device_ids);

    match devices.add(&sync_db.0.get_connection()) {
        Ok(_) => Some(true),
        Err(e) => {
            eprintln!("[Device] Failed to add device! {}", e);
            Some(false)
        }
    }
}

#[tauri::command]
pub fn list_devices(sync_db: State<ConfigDatabase>) -> Option<Vec<String>> {
    match Device::list(&sync_db.0.get_connection()) {
        Ok(ids) => Some(ids),
        Err(e) => {
            eprintln!("[Device] Failed to fetch! {}", e);
            None
        }
    }
}
